File: Packages/Core/Sources/FissionCore/Domains/Thread/ProjectPathQuery.swift
import Foundation

/// Path-shaped project search used by the New Thread picker.
public enum ProjectPathQuery {
    public static func isPathQuery(_ query: String) -> Bool {
        query.hasPrefix("./") || query.hasPrefix("~/") || query.hasPrefix("/")
    }

    /// Remote queries without `./`, `~/`, or `/` are looked up under `~/`.
    public static func normalizeRemoteQuery(_ query: String) -> String {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty || trimmed == "~" { return "~/" }
        if isPathQuery(trimmed) { return trimmed }
        return "~/" + trimmed
    }

    /// Directory to list and the filename prefix to match.
    public static func listingTarget(
        query: String,
        relativeBase: String?
    ) -> (directory: String, namePrefix: String)? {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard isPathQuery(trimmed) else { return nil }

        let expanded = expandRelative(trimmed, base: relativeBase)
        if trimmed.hasSuffix("/") {
            return (stripTrailingSlashes(expanded), "")
        }
        return (parentDirectory(expanded), lastComponent(expanded))
    }

    /// If the typed last component is an exact directory name, list inside it.
    public static func resolvedListingTarget(
        query: String,
        relativeBase: String?,
        exactChildNames: ((String) -> [String]?)? = nil
    ) -> (directory: String, namePrefix: String)? {
        guard let target = listingTarget(query: query, relativeBase: relativeBase) else {
            return nil
        }
        guard !target.namePrefix.isEmpty,
              let names = exactChildNames?(target.directory),
              let exact = names.first(where: {
                  $0.compare(target.namePrefix, options: [.caseInsensitive]) == .orderedSame
              }) else {
            return target
        }
        return (join(target.directory, exact), "")
    }

    /// Children at `directory`, plus the directory itself when browsing inside it.
    public static func pickerPaths(
        directory: String,
        namePrefix: String,
        childNames: [String]
    ) -> [String] {
        let children = childProjects(
            directory: directory,
            names: childNames,
            namePrefix: namePrefix
        )
        guard namePrefix.isEmpty,
              let current = MoshCommand.normalizedDirectory(directory),
              current != "/",
              current != "~",
              current != "." else {
            return children
        }
        return [current] + children.filter { $0 != current }
    }

    public static func expandRelative(_ path: String, base: String?) -> String {
        guard path.hasPrefix("./") else { return path }
        let rest = String(path.dropFirst(2))
        guard let base, !base.isEmpty else { return path }
        if rest.isEmpty || rest == "/" {
            return stripTrailingSlashes(base)
        }
        return join(stripTrailingSlashes(base), rest)
    }

    public static func join(_ directory: String, _ child: String) -> String {
        if child.isEmpty { return directory }
        if directory == "/" {
            return child.hasPrefix("/") ? child : "/" + child
        }
        if directory.hasSuffix("/") {
            return directory + child
        }
        return directory + "/" + child
    }

    public static func parentDirectory(_ path: String) -> String {
        let trimmed = stripTrailingSlashes(path)
        if trimmed == "/" || trimmed == "~" || trimmed == "." { return trimmed }
        guard let slash = trimmed.lastIndex(of: "/") else { return "." }
        let parent = String(trimmed[..<slash])
        if parent.isEmpty { return "/" }
        return parent
    }

    public static func lastComponent(_ path: String) -> String {
        let trimmed = stripTrailingSlashes(path)
        if trimmed == "/" { return "/" }
        if trimmed == "~" { return "~" }
        guard let slash = trimmed.lastIndex(of: "/") else { return trimmed }
        let name = String(trimmed[trimmed.index(after: slash)...])
        return name.isEmpty ? trimmed : name
    }

    public static func matchesPrefix(_ name: String, prefix: String) -> Bool {
        prefix.isEmpty
            || name.range(of: prefix, options: [.caseInsensitive, .anchored]) != nil
    }

    public static func matchesRecent(_ path: String, query: String) -> Bool {
        query.isEmpty
            || lastComponent(path).localizedCaseInsensitiveContains(query)
            || path.localizedCaseInsensitiveContains(query)
    }

    /// Tab-complete a selected path, keeping `./` and `~/` prefixes from the query.
    public static func completedInput(
        query: String,
        path: String,
        homePath: String?,
        relativeBase: String?
    ) -> String {
        let withSlash = path.hasSuffix("/") || path == "/" ? path : path + "/"

        if query.hasPrefix("./"), let relativeBase {
            let base = stripTrailingSlashes(relativeBase)
            if path == base { return "./" }
            let prefix = base + "/"
            if path.hasPrefix(prefix) {
                return "./" + path.dropFirst(prefix.count) + (path.hasSuffix("/") ? "" : "/")
            }
        }

        if let homePath {
            let home = stripTrailingSlashes(homePath)
            if query.hasPrefix("~/") || (!query.hasPrefix("/") && path.hasPrefix(home + "/"))
                || path == home
            {
                if path == home { return "~/" }
                if path.hasPrefix(home) {
                    return "~" + path.dropFirst(home.count) + (path.hasSuffix("/") ? "" : "/")
                }
            }
        }

        if path.hasPrefix("~") {
            return withSlash
        }

        return withSlash
    }

    public static func recentProjects(
        _ paths: [String],
        matching query: String,
        limit: Int = 9
    ) -> [String] {
        var seen = Set<String>()
        var result: [String] = []
        for path in paths {
            guard let normalized = MoshCommand.normalizedDirectory(path) else { continue }
            guard seen.insert(normalized).inserted else { continue }
            guard matchesRecent(normalized, query: query) else { continue }
            result.append(normalized)
            if result.count == limit { break }
        }
        return result
    }

    public static func childProjects(
        directory: String,
        names: [String],
        namePrefix: String,
        limit: Int = 50
    ) -> [String] {
        names
            .filter { matchesPrefix($0, prefix: namePrefix) }
            .sorted { $0.localizedCaseInsensitiveCompare($1) == .orderedAscending }
            .prefix(limit)
            .map { join(directory, $0) }
    }

    public static func stripTrailingSlashes(_ path: String) -> String {
        var trimmed = path
        while trimmed.count > 1, trimmed.hasSuffix("/") {
            trimmed.removeLast()
        }
        return trimmed
    }
}

File: packages/electron/src/shared/projectPathQuery.ts
import { posixQuote } from "./moshCommand";

export type ListingTarget = {
  directory: string;
  namePrefix: string;
};

export function isPathQuery(query: string): boolean {
  return query.startsWith("./") || query.startsWith("~/") || query.startsWith("/");
}

export function normalizeRemoteQuery(query: string): string {
  const trimmed = query.trim();
  if (trimmed.length === 0 || trimmed === "~") {
    return "~/";
  }
  if (isPathQuery(trimmed)) {
    return trimmed;
  }
  return `~/${trimmed}`;
}

export function listingTarget(query: string, relativeBase: string | null): ListingTarget | null {
  const trimmed = query.trim();
  if (!isPathQuery(trimmed)) {
    return null;
  }
  const expanded = expandRelative(trimmed, relativeBase);
  if (trimmed.endsWith("/")) {
    return { directory: stripTrailingSlashes(expanded), namePrefix: "" };
  }
  return { directory: parentDirectory(expanded), namePrefix: lastComponent(expanded) };
}

export function resolvedListingTarget(
  query: string,
  relativeBase: string | null,
  exactChildNames?: (directory: string) => string[] | null
): ListingTarget | null {
  const target = listingTarget(query, relativeBase);
  if (!target) {
    return null;
  }
  if (target.namePrefix.length === 0) {
    return target;
  }
  const names = exactChildNames?.(target.directory);
  const exact = names?.find(
    (name) => name.localeCompare(target.namePrefix, undefined, { sensitivity: "accent" }) === 0
  );
  if (!exact) {
    return target;
  }
  return { directory: join(target.directory, exact), namePrefix: "" };
}

export function pickerPaths(
  directory: string,
  namePrefix: string,
  childNames: string[],
  includeCurrentDirectory: boolean
): string[] {
  const children = childProjects(directory, childNames, namePrefix);
  const current = normalizedDirectory(directory);
  if (
    !includeCurrentDirectory ||
    namePrefix.length > 0 ||
    current == null ||
    current === "/" ||
    current === "~" ||
    current === "."
  ) {
    return children;
  }
  return [current, ...children.filter((path) => path !== current)];
}

export function expandRelative(path: string, base: string | null): string {
  if (!path.startsWith("./")) {
    return path;
  }
  const rest = path.slice(2);
  if (base == null || base.length === 0) {
    return path;
  }
  if (rest.length === 0 || rest === "/") {
    return stripTrailingSlashes(base);
  }
  return join(stripTrailingSlashes(base), rest);
}

export function join(directory: string, child: string): string {
  if (child.length === 0) {
    return directory;
  }
  if (directory === "/") {
    return child.startsWith("/") ? child : `/${child}`;
  }
  if (directory.endsWith("/")) {
    return directory + child;
  }
  return `${directory}/${child}`;
}

export function parentDirectory(path: string): string {
  const trimmed = stripTrailingSlashes(path);
  if (trimmed === "/" || trimmed === "~" || trimmed === ".") {
    return trimmed;
  }
  const slash = trimmed.lastIndexOf("/");
  if (slash === -1) {
    return ".";
  }
  const parent = trimmed.slice(0, slash);
  return parent.length === 0 ? "/" : parent;
}

export function lastComponent(path: string): string {
  const trimmed = stripTrailingSlashes(path);
  if (trimmed === "/") {
    return "/";
  }
  if (trimmed === "~") {
    return "~";
  }
  const slash = trimmed.lastIndexOf("/");
  if (slash === -1) {
    return trimmed;
  }
  const name = trimmed.slice(slash + 1);
  return name.length === 0 ? trimmed : name;
}

export function matchesPrefix(name: string, prefix: string): boolean {
  return prefix.length === 0 || name.toLowerCase().startsWith(prefix.toLowerCase());
}

export function matchesRecent(path: string, query: string): boolean {
  return (
    query.length === 0 ||
    lastComponent(path).toLowerCase().includes(query.toLowerCase()) ||
    path.toLowerCase().includes(query.toLowerCase())
  );
}

export function completedInput(
  query: string,
  path: string,
  homePath: string | null,
  relativeBase: string | null
): string {
  const withSlash = path.endsWith("/") || path === "/" ? path : `${path}/`;

  if (query.startsWith("./") && relativeBase) {
    const base = stripTrailingSlashes(relativeBase);
    if (path === base) {
      return "./";
    }
    const prefix = `${base}/`;
    if (path.startsWith(prefix)) {
      return `./${path.slice(prefix.length)}${path.endsWith("/") ? "" : "/"}`;
    }
  }

  if (homePath) {
    const home = stripTrailingSlashes(homePath);
    if (
      query.startsWith("~/") ||
      (!query.startsWith("/") && path.startsWith(`${home}/`)) ||
      path === home
    ) {
      if (path === home) {
        return "~/";
      }
      if (path.startsWith(home)) {
        return `~${path.slice(home.length)}${path.endsWith("/") ? "" : "/"}`;
      }
    }
  }

  if (path.startsWith("~")) {
    return withSlash;
  }

  return withSlash;
}

export function recentProjects(paths: string[], matching: string, limit = 9): string[] {
  const seen = new Set<string>();
  const result: string[] = [];
  for (const path of paths) {
    const normalized = normalizedDirectory(path);
    if (normalized == null || seen.has(normalized)) {
      continue;
    }
    seen.add(normalized);
    if (!matchesRecent(normalized, matching)) {
      continue;
    }
    result.push(normalized);
    if (result.length === limit) {
      break;
    }
  }
  return result;
}

export function childProjects(
  directory: string,
  names: string[],
  namePrefix: string,
  limit = 50
): string[] {
  return names
    .filter((name) => matchesPrefix(name, namePrefix))
    .sort((left, right) => left.localeCompare(right, undefined, { sensitivity: "accent" }))
    .slice(0, limit)
    .map((name) => join(directory, name));
}

export function stripTrailingSlashes(path: string): string {
  let trimmed = path;
  while (trimmed.length > 1 && trimmed.endsWith("/")) {
    trimmed = trimmed.slice(0, -1);
  }
  return trimmed;
}

export function normalizedDirectory(path: string | null | undefined): string | null {
  if (path == null) {
    return null;
  }
  let trimmed = path.trim();
  if (trimmed.length === 0) {
    return null;
  }
  while (trimmed.length > 1 && trimmed.endsWith("/")) {
    trimmed = trimmed.slice(0, -1);
  }
  return trimmed;
}

export function posixQuotePath(value: string): string {
  return posixQuote(value);
}
